### PR TITLE
rearrange and cleanup of build

### PR DESCRIPTION
**Pull request by John-Colvin**
Changes:
- folded benchmark building in to `dub.json` so it can automatically detect where e.g. emsi containers are.
- deleted `docs.json`, I presume it wasn't supposed to be checked in anyway?
- fold all of `docs.sh` in to `dub.json` for same reason as for benchmark builds
- `s/formated/formatted/` in `dub.json`
- move `library` build to first config so it is the default
- rearrange source code structure so as to have `package.d` working properly. Previously depending on date-parser and writing `import dateparser;` meant that dmd would look in `src` for either a file called `dateparser.d` or a folder called `dateparser` containing a `package.d`, neither of which would work.
- Now testing beta versions of compilers as well (and ldc 0.17.1)

`bench.sh` is still there, it just calls dub with the relevant flags

edit: accidentally wrote `package.json` instead of `package.d` above, sorry for the confusion

### CODE REVIEW

File: dub.json [hunk 1]
```diff
@@ -1,21 +1,40 @@
...
             "mainSourceFile": "bench.d",
             "targetType": "executable"
         },
         {
-            "name": "library",
-            "targetType": "library"
+            "name": "benchmark",
+            "mainSourceFile": "bench.d",
+            "targetType": "executable",
+            "dflags-ldc": ["-O5", "-singleobj", "-boundscheck=off"]
```
Comment (JackStouffer) [inline dub.json:23]: I think you want O3 here.

Comment (John-Colvin) [inline dub.json:23]: O5 is the highest optimisation level in ldc, IIRC currently it's just equivalent to O3 though

**Comment by John-Colvin** [comment]
woops, messed up the description, see edited version above ^

File: dub.json (hunk 1 above)
Comment (John-Colvin) [inline dub.json:23]: could also drop that entirely, as `dub` defaults to using `-O` for ldc release builds which is equivalent to `-O3`

**Comment by JackStouffer** [comment]
Add docs.json to the gitignore, then LGTM.

**Comment by John-Colvin** [comment]
done

**Pull request merged**